File: linearRegression.js
const CONFIG = require('./config');
const util = require('./util');

// const SLR = ml.SLR; // Simple Linear Regression
const MLR = require( 'ml-regression-multivariate-linear');


const DATASET = CONFIG.ACTIVE_LINEAR_REGRESSION_DATASET;
const TRAINING_PARAMS = CONFIG.TRAINING_PARAMS;





/*
  Applies linear model to test features
*/
function predictY(xTest, weights) {
  let predictions = [];
  for(var row=0; row<xTest.length; row++) {
    var sum = 0;
    for(var col=0; col<xTest[0].length; col++) {
      sum += xTest[row][col] * weights[col][0];
    }
    predictions.push(sum);
  }

  return predictions;
}

/*
  Calculates the Mean Square and Absolute Errors (MSE, MAE) 
*/
function evaluateModel(yTest, predictions) {
  let mse = [];
  let mae = [];
  const numFeautres = yTest.length;

  // Calculate individual metrics for each test case
  for(var i=0; i<numFeautres; i++) {
    var difference = yTest[i] - predictions[i];
    mae.push(Math.abs(difference));
    mse.push(Math.pow(difference, 2))
  }

  return {
    mse: util.getMeanOfVector(mse),
    mae: util.getMeanOfVector(mae)
  }
}

function train(xTrain, yTrain) {
  const mlr = new MLR(xTrain,yTrain);
  const weights = mlr.toJSON().weights;

  return weights;
}

function evaluate() {

}



// Main function starts here
util.readFromCsv.then((readData) => {
  let X = util.createDesignMatrix(readData);
  let y = util.createLabelVector(readData);
  
  X = util.standardizeData(X);
  y = util.standardizeData(y);

  let averageMSE, averageMAE;
  if(TRAINING_PARAMS['SPLIT_METHOD'] === 'KFOLD') {
    const kFoldMSE = [], kFoldMAE = [];
    for(let i=0; i<TRAINING_PARAMS['NUM_SPLITS']; i++) {
      const { xTrain, yTrain, xTest, yTest } = util.splitKFoldCrossVal(X, y, TRAINING_PARAMS['NUM_SPLITS'], i);
      console.log("Evaluation fold " + i + " - Training with X = [" + xTrain.length + "," + xTrain[0].length + "], y = [" + yTrain.length + "]");
      
      const weights = train(xTrain, yTrain);

      // Predict the test values
      const predictions = predictY(xTest, weights);

      // Evaluate the model
      const { mse, mae } = evaluateModel(yTest, predictions);
      kFoldMSE.push(mse);
      kFoldMAE.push(mae);
      console.log("Mean Square Error: ", mse);
      console.log("Mean Absolute Error: ", mae, "\n");
    }

    averageMSE = util.getMeanOfVector(kFoldMSE);
    averageMAE = util.getMeanOfVector(kFoldMAE);
  } else {
    const { xTrain, yTrain, xTest, yTest } = util.split7030(X, y);
    console.log("Evaluation with 70/30: training with X = [" + xTrain.length + "," + xTrain[0].length + "], y = [" + yTrain.length + "]");
  
    const weights = train(xTrain, yTrain);
    const predictions = predictY(xTest, weights);
    const evaluation = evaluateModel(yTest, predictions)
    averageMAE = evaluation.mae;
    averageMSE = evaluation.mse;
  }

  console.log("Average MSE: ", averageMSE);
  console.log("AverageMAE: ", averageMAE);
  

  // console.log("Starting training with X = [" + xTrain.length + "," + xTrain[0].length + "], y = [" + yTrain.length + "]");
  // const mlr = new MLR(xTrain,yTrain);
  // const weights = mlr.toJSON().weights;
  // console.log("Weights: ", weights);

  // // Predict the test values
  // const predictions = predictY(xTest, weights);

  // // Evaluate the model
  // const { mse, mae } = evaluateModel(yTest, predictions);
  // console.log("Mean Square Error: ", mse);
  // console.log("Mean Absolute Error: ", mae);

}).catch((err) => {
    console.log(err);
});

File: linear-regression.py
import util

import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt



# Prepends a column of ones to design matrix to represent bias term 
def prependBiasTerm(dataset):
  numInstances = dataset.shape[0]
  numFeatures = dataset.shape[1]
  X = np.reshape(np.c_[np.ones(numInstances),dataset],[numInstances,numFeatures + 1])
  return X


# Performs gradient descent on training data
def trainModel(xTrain, yTrain, sess, trainingStep, lossFunctionHistory):
    for epoch in range(util.TRAINING_PARAMS['TRAINING_EPOCHS']):
        sess.run(trainingStep,feed_dict={X:xTrain,Y:yTrain})
        lossFunctionHistory = np.append(lossFunctionHistory, sess.run(cost,feed_dict={X: xTrain,Y: yTrain}))

    return lossFunctionHistory

# Returns the Mean Square and Mean Absolute Erros
def evaluateModel(yTest, predictedY):
    mse = sess.run(tf.reduce_mean(tf.square(predictedY - yTest)))
    mae = sess.run(tf.reduce_mean(tf.abs(predictedY - yTest)))

    return mse, mae


 # Plot some of the training data against input features
def showPlots(X, y):
    for i in range(1,X.shape[1]):
        plt.plot(X[:,i], y, 'r+')
        plt.xlabel("Feature %d" % i)
        plt.ylabel(util.ACTIVE_DATASET['LABEL'])
        plt.show()



###########################################
####     Begin Data Organization    #######
###########################################

data = util.readData()
designMatrix = util.createDesignMatrix(data)
labelVector = util.createLabelVector(data)


designMatrix = util.featureNormalize(designMatrix)
labelVector = util.featureNormalize(labelVector)

designMatrix = prependBiasTerm(designMatrix)



###########################################
####     Model Definition           #######
###########################################
# Leaving as 'None' Rows so we can use for train and predict (different number of rows)
X = tf.placeholder(tf.float32, [None, designMatrix.shape[1]])
Y = tf.placeholder(tf.float32,[None,1])
W = tf.Variable(tf.ones([designMatrix.shape[1], 1]))
lossFunctionHistory = np.empty(shape=[1],dtype=float)

# Specify Linear Regression method and parameters
init = tf.global_variables_initializer()
yPredictor = tf.matmul(X, W)
cost = tf.reduce_mean(tf.square(Y-yPredictor))
trainingStep = tf.train.GradientDescentOptimizer(util.TRAINING_PARAMS['LEARNING_RATE']).minimize(cost)


# Create a session
sess = tf.Session()
sess.run(init)


# Train the model
if(util.TRAINING_PARAMS['SPLIT_METHOD'] == "KFOLD"):        
    mse = []
    mae = []
    numDataSplits = util.TRAINING_PARAMS['NUM_SPLITS']
    for i in range(numDataSplits):
        xTrain, yTrain, xTest, yTest = util.splitUpDataCrossVal(designMatrix, labelVector, numDataSplits, crossValIndex=i)
        trainModel(xTrain, yTrain, sess, trainingStep, lossFunctionHistory)
        predictedY = sess.run(yPredictor, feed_dict={X: xTest})
        currentMSE, currentMAE = evaluateModel(yTest, predictedY)
        mse.append(currentMSE)
        mae.append(currentMAE)
    print("MSE: ", mse)
    print("MAE: ", mae)
    averageMSE = np.mean(mse)
    averageMAE = np.mean(mae)

else:
    xTrain, yTrain, xTest, yTest = util.splitData7030(designMatrix, labelVector)
    trainModel(xTrain, yTrain, sess, trainingStep, lossFunctionHistory)
    predictedY = sess.run(yPredictor, feed_dict={X: xTest})
    averageMSE, averageMAE = evaluateModel(yTest, predictedY)

print("Most recent weights: ", np.matrix(sess.run(W)))
print("Average MSE: %4f" % averageMSE, ", Average MAE: %4f" % averageMAE)





###########################################
####         Evaluation             #######
###########################################
# # Plot the cost function over time
# plt.plot(range(len(lossFunctionHistory)), lossFunctionHistory, 'b+')
# plt.xlabel("Epoch #")
# plt.ylabel("Cost function")
# plt.axis([0,util.TRAINING_PARAMS['TRAINING_EPOCHS'],0,np.max(lossFunctionHistory) + (0.1*np.max(lossFunctionHistory))])
# plt.show()


# Plot first feature against predicter / measured values
plt.plot(xTest[:,1], yTest, 'ro')
plt.plot(xTest[:,1], predictedY, 'bx')
plt.xlabel("Feature 1")
plt.ylabel(util.ACTIVE_DATASET['LABEL'])
plt.show()

# # Plot Predicted y values against measured y values
plt.plot(predictedY, yTest, 'ro')

if((util.TRAINING_PARAMS['NORMALIZE_METHOD'] == "MINMAX")):
    # Plot over range from [0,1]
    plt.plot([0, 1])
else:
    # Plot over range [-3, 3] ~ 99 % of data
    plt.plot(range(-3,3), range(-3,3), 'b')
plt.xlabel("Predicted")
plt.ylabel("Actual")
plt.show()